// @capacitor-community/text-to-speech/dist/plugin.cjs.js
'use strict';

var core = require('@capacitor/core');

exports.QueueStrategy = void 0;
(function (QueueStrategy) {
    /**
     * Use `Flush` to stop the current request when a new request is sent.
     */
    QueueStrategy[QueueStrategy["Flush"] = 0] = "Flush";
    /**
     * Use `Add` to buffer the speech request. The request will be executed when all previous requests have been completed.
     */
    QueueStrategy[QueueStrategy["Add"] = 1] = "Add";
})(exports.QueueStrategy || (exports.QueueStrategy = {}));

const TextToSpeech = core.registerPlugin('TextToSpeech', {
    web: () => Promise.resolve().then(function () { return web; }).then((m) => new m.TextToSpeechWeb()),
});

class TextToSpeechWeb extends core.WebPlugin {
    constructor() {
        super();
        this.speechSynthesis = null;
        if ('speechSynthesis' in window) {
            this.speechSynthesis = window.speechSynthesis;
            window.addEventListener('beforeunload', () => {
                this.stop();
            });
        }
    }
    async speak(options) {
        if (!this.speechSynthesis) {
            this.throwUnsupportedError();
        }
        await this.stop();
        const speechSynthesis = this.speechSynthesis;
        const utterance = this.createSpeechSynthesisUtterance(options);
        return new Promise((resolve, reject) => {
            utterance.onend = () => {
                resolve();
            };
            utterance.onerror = (event) => {
                reject(event);
            };
            speechSynthesis.speak(utterance);
        });
    }
    async stop() {
        if (!this.speechSynthesis) {
            this.throwUnsupportedError();
        }
        this.speechSynthesis.cancel();
    }
    async getSupportedLanguages() {
        const voices = this.getSpeechSynthesisVoices();
        const languages = voices.map((voice) => voice.lang);
        const filteredLanguages = languages.filter((v, i, a) => a.indexOf(v) == i);
        return { languages: filteredLanguages };
    }
    async getSupportedVoices() {
        const voices = this.getSpeechSynthesisVoices();
        return { voices };
    }
    async isLanguageSupported(options) {
        const result = await this.getSupportedLanguages();
        const isLanguageSupported = result.languages.includes(options.lang);
        return { supported: isLanguageSupported };
    }
    async openInstall() {
        this.throwUnimplementedError();
    }
    createSpeechSynthesisUtterance(options) {
        const voices = this.getSpeechSynthesisVoices();
        const utterance = new SpeechSynthesisUtterance();
        const { text, lang, rate, pitch, volume, voice } = options;
        // Compare against undefined so an explicit 0 (a valid voice index, volume, or pitch) is applied.
        if (voice !== undefined) {
            utterance.voice = voices[voice];
        }
        if (volume !== undefined) {
            utterance.volume = volume >= 0 && volume <= 1 ? volume : 1;
        }
        if (rate !== undefined) {
            utterance.rate = rate >= 0.1 && rate <= 10 ? rate : 1;
        }
        if (pitch !== undefined) {
            utterance.pitch = pitch >= 0 && pitch <= 2 ? pitch : 2;
        }
        if (lang) {
            utterance.lang = lang;
        }
        utterance.text = text;
        return utterance;
    }
    getSpeechSynthesisVoices() {
        if (!this.speechSynthesis) {
            this.throwUnsupportedError();
        }
        if (!this.supportedVoices || this.supportedVoices.length < 1) {
            this.supportedVoices = this.speechSynthesis.getVoices();
        }
        return this.supportedVoices;
    }
    throwUnsupportedError() {
        throw this.unavailable('SpeechSynthesis API not available in this browser.');
    }
    throwUnimplementedError() {
        throw this.unimplemented('Not implemented on web.');
    }
}

var web = /*#__PURE__*/Object.freeze({
    __proto__: null,
    TextToSpeechWeb: TextToSpeechWeb
});

exports.TextToSpeech = TextToSpeech;
